File: apps/api/src/db/schedule.py
from enum import Enum
from typing import Optional
from sqlalchemy import Column, ForeignKey, Integer
from sqlmodel import Field, SQLModel

from src.db.users import UserRead


class ScheduleSessionStatus(str, Enum):
    SCHEDULED = "scheduled"
    CANCELLED = "cancelled"
    COMPLETED = "completed"
    NO_SHOW = "no_show"


class ScheduleNotificationType(str, Enum):
    SESSION_BOOKED = "session_booked"
    SESSION_CANCELLED = "session_cancelled"


class TutorAssignmentBase(SQLModel):
    org_id: int = Field(sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), index=True))
    tutor_user_id: int = Field(sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True))
    student_user_id: int = Field(sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True))
    course_id: Optional[int] = Field(default=None, foreign_key="course.id")
    active: bool = True


class TutorAssignment(TutorAssignmentBase, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    assignment_uuid: str = Field(default="", index=True)
    created_by_id: int = Field(default=0, foreign_key="user.id")
    creation_date: str = ""
    update_date: str = ""


class TutorAssignmentCreate(SQLModel):
    tutor_user_id: int
    student_user_id: int
    course_id: Optional[int] = None
    active: bool = True


class TutorAssignmentRead(TutorAssignmentBase):
    id: int
    assignment_uuid: str
    tutor: UserRead
    student: UserRead
    created_by_id: int
    creation_date: str
    update_date: str


class TutorAvailabilityBase(SQLModel):
    org_id: int = Field(sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), index=True))
    tutor_user_id: int = Field(sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True))
    weekday: int = Field(ge=0, le=6)
    start_time: str
    end_time: str
    slot_minutes: int = 30
    timezone: str = "UTC"
    active: bool = True


class TutorAvailability(TutorAvailabilityBase, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    availability_uuid: str = Field(default="", index=True)
    creation_date: str = ""
    update_date: str = ""


class TutorAvailabilityCreate(SQLModel):
    tutor_user_id: int
    weekday: int
    start_time: str
    end_time: str
    slot_minutes: int = 30
    timezone: str = "UTC"
    active: bool = True


class TutorAvailabilityRead(TutorAvailabilityBase):
    id: int
    availability_uuid: str
    creation_date: str
    update_date: str


class ScheduleSessionBase(SQLModel):
    org_id: int = Field(sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), index=True))
    tutor_user_id: int = Field(sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True))
    student_user_id: int = Field(sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True))
    course_id: Optional[int] = Field(default=None, foreign_key="course.id")
    starts_at: str = Field(index=True)
    ends_at: str = Field(index=True)
    timezone: str = "UTC"
    student_notes: Optional[str] = ""
    status: ScheduleSessionStatus = ScheduleSessionStatus.SCHEDULED


class ScheduleSession(ScheduleSessionBase, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    session_uuid: str = Field(default="", index=True)
    created_by_id: int = Field(default=0, foreign_key="user.id")
    cancellation_reason: Optional[str] = ""
    creation_date: str = ""
    update_date: str = ""


class ScheduleSessionCreate(SQLModel):
    tutor_user_id: int
    starts_at: str
    ends_at: str
    timezone: str = "UTC"
    student_notes: Optional[str] = ""
    course_id: Optional[int] = None


class ScheduleSessionRead(ScheduleSessionBase):
    id: int
    session_uuid: str
    tutor: UserRead
    student: UserRead
    created_by_id: int
    cancellation_reason: Optional[str] = ""
    creation_date: str
    update_date: str


class ScheduleNotification(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    notification_uuid: str = Field(default="", index=True)
    org_id: int = Field(sa_column=Column(Integer, ForeignKey("organization.id", ondelete="CASCADE"), index=True))
    user_id: int = Field(sa_column=Column(Integer, ForeignKey("user.id", ondelete="CASCADE"), index=True))
    type: ScheduleNotificationType
    title: str
    body: str
    session_uuid: Optional[str] = Field(default="", index=True)
    read_at: Optional[str] = None
    creation_date: str = ""


class ScheduleNotificationRead(SQLModel):
    id: int
    notification_uuid: str
    org_id: int
    user_id: int
    type: ScheduleNotificationType
    title: str
    body: str
    session_uuid: Optional[str] = ""
    read_at: Optional[str] = None
    creation_date: str


class ScheduleSummary(SQLModel):
    is_admin: bool
    is_tutor: bool
    assigned_tutors: list[UserRead]
    upcoming_sessions: list[ScheduleSessionRead]
    notifications: list[ScheduleNotificationRead]


class ScheduleSlot(SQLModel):
    starts_at: str
    ends_at: str
    available: bool

File: apps/web/services/schedule/schedule.ts
import { getAPIUrl } from '@services/config/config'
import {
  RequestBodyWithAuthHeader,
  errorHandling,
} from '@services/utils/ts/requests'

export type ScheduleUser = {
  id: number
  user_uuid: string
  username: string
  first_name?: string
  last_name?: string
  email: string
  avatar_image?: string
  details?: any
  bio?: string
}

export type ScheduleSession = {
  id: number
  session_uuid: string
  tutor_user_id: number
  student_user_id: number
  tutor: ScheduleUser
  student: ScheduleUser
  starts_at: string
  ends_at: string
  timezone: string
  student_notes?: string
  status: 'scheduled' | 'cancelled' | 'completed' | 'no_show'
  status_marked_by?: ScheduleUser | null
  status_marked_by_id?: number | null
  status_marked_at?: string | null
  instructor_notes?: string
  cancellation_reason?: string
}

export type ScheduleSummary = {
  is_admin: boolean
  is_tutor: boolean
  assigned_tutors: ScheduleUser[]
  upcoming_sessions: ScheduleSession[]
  notifications: Array<{
    id: number
    notification_uuid: string
    title: string
    body: string
    session_uuid?: string
    read_at?: string
    creation_date: string
  }>
}

export type ScheduleSlot = {
  starts_at: string
  ends_at: string
  available: boolean
}

export type TutorAvailability = {
  id: number
  availability_uuid: string
  org_id: number
  tutor_user_id: number
  weekday: number
  start_time: string
  end_time: string
  slot_minutes: number
  timezone: string
  active: boolean
  creation_date: string
  update_date: string
}

export async function getScheduleSummary(orgId: number, accessToken?: string) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/summary`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result)
}

export async function getTutorSlots(
  orgId: number,
  tutorUserId: number,
  fromDate: string,
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/tutors/${tutorUserId}/slots?from_date=${fromDate}&days=7`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result)
}

export async function getScheduleSessions(orgId: number, accessToken?: string) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/sessions`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result)
}

export async function createScheduleSession(
  orgId: number,
  data: {
    tutor_user_id: number
    starts_at: string
    ends_at: string
    timezone: string
    student_notes?: string
  },
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/sessions`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return errorHandling(result)
}

export async function cancelScheduleSession(
  orgId: number,
  sessionUuid: string,
  reason: string,
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/sessions/${sessionUuid}/cancel`,
    RequestBodyWithAuthHeader('POST', { reason }, null, accessToken)
  )
  return errorHandling(result)
}

export async function updateScheduleSessionStatus(
  orgId: number,
  sessionUuid: string,
  data: {
    status: 'scheduled' | 'completed' | 'no_show'
    instructor_notes?: string
  },
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/sessions/${sessionUuid}/status`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return errorHandling(result)
}

export async function getTutorAvailability(
  orgId: number,
  tutorUserId: number,
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/tutors/${tutorUserId}/availability`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result)
}

export async function createTutorAvailability(
  orgId: number,
  data: {
    tutor_user_id: number
    weekday: number
    start_time: string
    end_time: string
    slot_minutes: number
    timezone: string
  },
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/availability`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return errorHandling(result)
}

export async function updateTutorAvailability(
  orgId: number,
  availabilityUuid: string,
  data: {
    weekday: number
    start_time: string
    end_time: string
    slot_minutes: number
    timezone: string
    active?: boolean
  },
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/availability/${availabilityUuid}`,
    RequestBodyWithAuthHeader('PATCH', data, null, accessToken)
  )
  return errorHandling(result)
}

export async function deleteTutorAvailability(
  orgId: number,
  availabilityUuid: string,
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/availability/${availabilityUuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, accessToken)
  )
  return errorHandling(result)
}

export async function createTutorAssignment(
  orgId: number,
  data: {
    tutor_user_id: number
    student_user_id: number
    course_id?: number | null
    active?: boolean
  },
  accessToken?: string
) {
  const result = await fetch(
    `${getAPIUrl()}schedule/org/${orgId}/assignments`,
    RequestBodyWithAuthHeader('POST', data, null, accessToken)
  )
  return errorHandling(result)
}

export async function getOrgUsersForSchedule(orgId: number, accessToken?: string) {
  const result = await fetch(
    `${getAPIUrl()}orgs/${orgId}/users?page=1&limit=100`,
    RequestBodyWithAuthHeader('GET', null, null, accessToken)
  )
  return errorHandling(result)
}
